torch/distributed/optim/named_optimizer.py

```diff
@@ -0,0 +1,244 @@
+        self.param_groups: Collection[Mapping[str, Any]] = param_groups  # type: ignore[assignment]
+        self.named_parameters = dict(named_parameters)
+        params_optimizer = (
+            self.named_parameters.values() if param_groups is None else param_groups
+        )
+        self._optimizer = optimizer_class(  # type: ignore[operator]
+            params_optimizer,
+            *args,
+            **kwargs,
+        )
+        if param_groups is None:
+            self.param_keys_order = list(self.named_parameters.keys())
+        else:
+            warnings.warn(
+                "Since we pass in param_groups, we will use param_groups to "
+                "initialize the optimizer, not all parameters of the module."
+            )
+            param_to_key = {param: key for key, param in self.named_parameters}  # type: ignore[misc, has-type]
+            param_keys_order = []
+            for group in param_groups:
+                for param in group["params"]:
+                    if param not in param_to_key:
+                        raise ValueError(f"Expect param name {param} found in param group but is missing.")
+                    param_keys_order.append(param_to_key[param])
+            self.param_keys_order = param_keys_order
+
+    def state_dict(self) -> Dict[str, Any]:
+        """
+        Return the state_dict of the optimzer. Instead of using number to index
+        parameters, we will use module fully qualifed name (FQN) as the key.
+        """
+        state_dict = self._optimizer.state_dict()
+        param_groups = state_dict["param_groups"]
+
+        ret_state = {
+            self.param_keys_order[st_key]: state_val
+            for st_key, state_val in state_dict["state"].items()
+        }
+
+        ret_groups = []
+        for group in param_groups:
+            param_keys = []
+            for param in group["params"]:
+                param_keys.append(self.param_keys_order[param])
+            ret_group = {"params": sorted(param_keys)}
+            for k, v in group.items():
+                if k != "params":
+                    ret_group[k] = deepcopy(v)
+            ret_groups.append(ret_group)
+
+        return {"state": ret_state, "param_groups": ret_groups}
+
+    def register_state_dict_pre(self):
+        raise NotImplementedError("register_state_dict_pre not supported yet and might be implemented soon.")
+
+    def register_state_dict_post(self):
+        raise NotImplementedError("register_state_dict_post not supported yet and might be implemented soon.")
+
+    def step(self):
+        """
+        Performs a single optimization step.
+
+        This will call :meth:`torch.optim.Optimizer.step` on the wrapped
+        optimizer.
+        """
+        self._optimizer.step()
+
+    def load_state_dict(self, state_dict: Mapping[str, Any]) -> None:
+        """
+        This public function defines the default behavior to load a state_dict
+        for named optimizer.
+
```

Comment:
For a follow-up: We should make sure this sample code renders as expected. The syntax looks a bit different than other examples I have seen.

Reply:
Sure, where do you check the render result?

Reply:
I think you have to add `NamedOptimizer` to the `distributed.optim.rst` file, but I think the class needs to be public (meaning that if you change it to `_NamedOptimizer`, then it will not render).

https://github.com/pytorch/pytorch/blob/master/docs/source/distributed.optim.rst

Feel free to refer to `fsdp.rst` as an example:
https://github.com/pytorch/pytorch/blob/master/docs/source/fsdp.rst

Reply:
I see. Thanks. They did mentioned it in https://github.com/pytorch/pytorch/blob/master/docs/source/community/contribution_guide.rst.